sfxclient/multitokensink.go

```diff
@@ -0,0 +1,734 @@
+						break outer
+					}
+				}
+			case <-a.requestDatapoints:
+				dp := []*datapoint.Datapoint{Cumulative(a.name, a.defaultDims, a.val)}
+				a.previousVal = dp
+				a.returnDatapoints <- dp
+			}
+		}
+	}()
+	return a
+}
+
+type asyncTokenStatusMap struct {
+	name              string
+	dataStore         map[string]map[string]int64
+	input             chan *tokenStatus
+	stop              chan bool
+	requestDatapoints chan bool
+	returnDatapoints  chan []*datapoint.Datapoint
+	previousVal       []*datapoint.Datapoint
+	defaultDims       map[string]string
+}
+
+func (a *asyncTokenStatusMap) fetchDatapoints() (counters []*datapoint.Datapoint) {
+	for token, statuses := range a.dataStore {
+		for status, counter := range statuses {
+			dims := map[string]string{"token": token, "status": status}
+			for k, v := range a.defaultDims {
+				dims[k] = v
+			}
+			counters = append(counters, Cumulative(a.name, dims, counter))
+		}
+	}
+	a.previousVal = counters
+	return
+}
+
+func (a *asyncTokenStatusMap) processInput(t *tokenStatus) {
+	if _, ok := a.dataStore[t.token]; !ok {
+		a.dataStore[t.token] = map[string]int64{}
+	}
+	if val, ok := a.dataStore[t.token][t.status]; ok {
+		a.dataStore[t.token][t.status] = val + t.val
+	} else { // if the status doesn't exist add create it
+		a.dataStore[t.token][t.status] = t.val
+	}
+}
+
+func (a *asyncTokenStatusMap) Datapoints() (dps []*datapoint.Datapoint) {
+	var request bool
+	for {
+		select {
+		case dps = <-a.returnDatapoints:
+			if dps == nil {
+				dps = a.previousVal
+			}
+			return
+		default:
+			if !request {
+				a.requestDatapoints <- true
+				request = true
+			}
+		}
+	}
+}
+
+// AddDatapoints adds a tokenStatus object to the counter
+func (a *asyncTokenStatusMap) Increment(status *tokenStatus) {
+	select {
+	case <-a.stop: // check if the counter has been stopped
+		return
+	default:
+		select {
+		case a.input <- status:
+		default:
+		}
+	}
+}
+
+func newAsyncTokenStatusMap(name string, buffer int, numWorkers int64, defaultDims map[string]string) *asyncTokenStatusMap {
+	a := &asyncTokenStatusMap{
+		name:              name,
+		dataStore:         map[string]map[string]int64{},
+		input:             make(chan *tokenStatus, int64(buffer)*numWorkers),
+		stop:              make(chan bool),
+		requestDatapoints: make(chan bool, 5000),
+		returnDatapoints:  make(chan []*datapoint.Datapoint, 5000),
+		defaultDims:       defaultDims,
+	}
+	go func() {
+		for {
+			select {
+			case <-a.stop: // signal for the goroutine to stop managing the datapoints object
+				a.fetchDatapoints() // fetch datapoints one last time and set it to previousVal so anyone reading can get it
+				close(a.returnDatapoints)
+				return
+			case input := <-a.input:
+				a.processInput(input)
+			outer:
+				for {
+					select {
+					case input := <-a.input:
+						a.processInput(input)
+					default:
+						break outer
+					}
+				}
+			case <-a.requestDatapoints:
+				a.returnDatapoints <- a.fetchDatapoints()
```

Comment:
i'd put the channel for the returned datapoints in the request object.  It means that this model can be extended to multiple clients (not specifically necessary here, but a good practice in general i think).